### DISCUSSION TITLE
Tokens are not refilled fully at Redis Cluster mode

### DISCUSSION THREAD

**Discussion by NaayoungKwon** [vladimir-bukhtoyarov/bucket4j#477] [Q&A] [answered]
Hello.
I'm trying to figure out problem about refilling tokens.

my environment is this.
- bucket4j 8.7.0
- use AWS Elastic Cache cluster mode
- use LettuceBasedProxyManager
- distributed system

```java
LettuceBasedProxyManager<byte[]> proxyManager = LettuceBasedProxyManager LettuceBasedProxyManager.builderFor(redisClusterClient)
    .withExpirationStrategy(ExpirationAfterWriteStrategy.basedOnTimeForRefillingBucketUpToMax(
            Duration.ofSeconds(1)))
        .build();

String key = "bucket-key";
Bandwidth bandwidth = getCustomBandwidth(key); // choose bandwith
Bucket bucket = proxyManager.builder().build(key.getBytes(), () -> BucketConfiguration.builder().addLimit(bandwidth).build());

ConsumptionProbe probe = bucket.tryConsumeAndReturnRemaining(1);
canConsume = probe.isConsumed();
log.info("RateLimitFilter|key:{}|remain:{}|waitToRefill:{}ms", key, probe.getRemainingTokens(), probe.getNanosToWaitForRefill() / 1_000_000);          

```
```java
public enum CustomBandwidth {

  B_DEFAULT {
    @Override
    public Bandwidth getLimit() {
      return Bandwidth.builder().capacity(100).refillIntervally(100, Duration.ofSeconds(1)).build();
    }
  },
  B_200 {
    @Override
    public Bandwidth getLimit() {
      return Bandwidth.builder().capacity(200).refillIntervally(200, Duration.ofSeconds(1)).build();
    }
  },
  ;

  public abstract Bandwidth getLimit();

}
```

Even though choosing B_200 which refill 200 for every second, printed log shows biggest remaining token is not 199 but 99.

```
RateLimitFilter|key:A|remain:0|waitToRefill:10ms
RateLimitFilter|key:A|remain:0|waitToRefill:7ms
RateLimitFilter|key:A|remain:0|waitToRefill:0ms
RateLimitFilter|key:A|remain:99|waitToRefill:0ms
RateLimitFilter|key:A|remain:98|waitToRefill:0ms
```

This situation doesn't happen in my local environment and it only happen specific key.
(key#1, key#2 use B_200 but only key#1 happend)
```
RateLimitFilter|key:B|remain:0|waitToRefill:10ms
RateLimitFilter|key:B|remain:0|waitToRefill:7ms
RateLimitFilter|key:B|remain:0|waitToRefill:0ms
RateLimitFilter|key:B|remain:199|waitToRefill:0ms
RateLimitFilter|key:B|remain:198|waitToRefill:0ms
```

I guess it would be related with high request rate.

If my configuration is wrong, please let me know 😭

**Comment by vladimir-bukhtoyarov** [comment, marked as answer]
Hello @NaayoungKwon 

> I guess it would be related with high request rate.

I do not think so. 

![image](https://github.com/bucket4j/bucket4j/assets/6761830/0eaa8854-00fe-41ee-a911-ffdb7f67e347)

As mentioneed in Java docs, configuration supplier is not called if bucket is already persisted. You continue to observe capacity 100, that means that bucket was already persisted, when you deployed new code that uses new configs, and something activity prolongs  lifetime of bucket that prevents bucket to be expired by TTL, so there is no chance for configuration supplier to be called. I suppose that previously you had only one configuration and splitted its later.

There is two right way to apply configuration changes:
* [Explicit configuration replacement](https://bucket4j.com/8.10.1/toc.html#configuration-replacement)
* [Implicit configuration replacement](https://bucket4j.com/8.10.1/toc.html#implicit-configuration-replacement)
I have lack information about your use-case, but I can suppose that second would be more preferable for you.

**Comment by NaayoungKwon** [comment]
Hello. @vladimir-bukhtoyarov 
Thanks for your answer.

the proxy manager is created as bean at spring and I choose bandwidth by requestor ip. 
so, this distribution system always use various key and bandwidth.

am I right to understand?
- ImplicitConfigurationReplacement : TokensInheritanceStrategy.RESET replace configuration every request and take current tokens that already saved from Redis.

I'll change like this.
```java
Bucket bucket = proxyManager.builder()
          .withImplicitConfigurationReplacement(1, RESET)
          .build(key.getBytes(), () -> BucketConfiguration.builder().addLimit(bandwidth).build());
```

Also, could you explain more about this? Does it mean, i have to change ExpirationAfterWriteStrategy?
>  I suppose that previously you had only one configuration and splitted its later.

**Reply by vladimir-bukhtoyarov** [reply]
> TokensInheritanceStrategy.RESET replace configuration every request

It reset bucket state in following cases:
* It detects that persisted configuration has version less than configured.
* Bucket was expired or bucket was never created before.

> Also, could you explain more about this? Does it mean, i have to change ExpirationAfterWriteStrategy?

Could you clarify your question? Why do you think that you need to change something in this?

**Reply by NaayoungKwon** [reply]
> > TokensInheritanceStrategy.RESET replace configuration every request
> 
> It reset bucket state in following cases:
> 
> * It detects that persisted configuration has version less than configured.
> * Bucket was expired or bucket was never created before.
> 

From this answer, I understand why you drew arrow from ExpirationAfterWriteStrategy at code.

![image](https://github.com/bucket4j/bucket4j/assets/34162358/bce32334-1c75-439a-ab14-e80ca3f8c5f1)

So, I understand like this.

1. for [key : #key1, Bandwidth : B_DEFAULT], get bucket A from Redis and consume
2. for [key : #key2, Bandwidth : B_200], get bucket B from Redis
    - bucket is different than before. so replace configuration to B_200.
3. bucket A's TTL expire
4. for [key : #key2, Bandwidth : B_200], get bucket B from Redis
    - bucket is same. so don't replace configuration.
5. for [key : #key1, Bandwidth : B_DEFAULT], bucket isn't exist. 
     - replace configuration to B_DEFAULT.

**Reply by vladimir-bukhtoyarov** [reply]
Yes, you understand correctly.

**Reply by NaayoungKwon** [reply]
I sincerely appreciate your help and support! It's really helpful.
